nullable-usage/README.md

```diff
@@ -0,0 +1,306 @@
+_NullPointerException_ which may ocuur at runtime.
+
+Also with the usage of nullable syntax, it is easier to visually see
+the intention of the code about the certaion issue ,
+
+To enable nullable in projects make sure you add the following configuration
+to your _.csproj_ file.
+
+```xml
+<Nullable>enable</Nullable>
+<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+```
+
+## Using Nullables in Mouseless
+
+We follow the guidelines below when writing code to properly use nullable syntax.
+
+### Don't use nullable for Dependency Injection
+
+```csharp
+public class PersonService
+{
+    readonly IQueryContext<Person>? _queryContext;
+ 
+    public PersonService(IQueryContext<Person>? queryContext)
+    {
+        _queryContext = queryContext;
+    }
+}
+```
+
+> :information_source:
+>
+> The DI container will resolve every dependency before initalizing the object
+> and an exception will be thrown for a if a component is not registered. 
+> Therefore, there is no possibility of dependency to be null.
+
+### Avoid using ! (null-forgiving) operator
+
+The usage of ! operator negates the null check that the compiler performs, so 
+the whole intention of enabling nullable check will be compromised and it
+will be very hard to spot possible bugs which may occur at runtime.
+
+! operator should only be used to dismiss the compiler errors when we set 
+default values for non nullable members which are;
+
+- Assigned with dependency injection
+- Assigned in the builder method
+
+```csharp
+public class Person
+{
+    readonly IEntityContext<Person> _context = default!;
+
+    public Person(IEntityContext<Person> context)
+    {
+        _context = context;
+    }
+
+    ...
+}
+```
+
+> :information_source:
+>
+> Because of NHibernate, entities need a _protected_ parameterless constructor 
+> and compiler will highlight an error stating that the value of __context_ is 
+> not assigned when leaving a constructor. Therefore we need to assign a 
+> default value to __context_ to dismiss the compiler error.  
+
+
+### Always manage nullable parameters
+
+When using nullable parameters, the main challenge is to manage what happens 
+within the scope of the method if a parameter has not been assigned with any 
+value. 
+
+We should always make sure the functionality is not broken and cause a wrong
+intention wether a parameter is necessary or not.
+
+#### Use not nullable when parameter is required
+
+```csharp
+public class Person
+{
+    public string Name { get; } = default!;
+    
+    ...
+
+    // name is made not nullable to give responsibility to caller
+    public virtual Person With(string name)
+    {
+        Name = name;
+    }
+}
+```
+
+### Handling null to not null situation
+
+#### Check for value and throw a relevant exception
+
+```csharp
+public class PersonService
+{
+    ...
+
+    public void AddPerson(string? name)
+    {
+        if(name is null) throw new ArgumentNullException();
+
+        _newPerson().With(name);
+    }
```

Comment:
although this example is technically correct, in practice it is unlikely in practice, thus making the example hard to understand.

a better real life example would be; a class implements an interface that accepts a nullable parameter, but has to pass it to a method that accepts not nullable parameter. in that case you would have no choice but make a null check at runtime.

this example is also hard to follow, i leave the final decision to you.